src/main/scala/nodes/learning/LBFGS.scala

```diff
@@ -0,0 +1,220 @@
+
+    /**
+      * NOTE: lossSum and loss is computed using the weights from the previous iteration
+      * and regVal is the regularization value computed in the previous iteration as well.
+      */
+    var state = states.next()
+    while (states.hasNext) {
+      lossHistory += state.value
+      state = states.next()
+    }
+    lossHistory += state.value
+    val finalWeights = state.x.asDenseMatrix.reshape(numFeatures, numClasses)
+
+    val lossHistoryArray = lossHistory.result()
+
+    logInfo("LBFGS.runLBFGS finished. Last 10 losses %s".format(
+      lossHistoryArray.takeRight(10).mkString(", ")))
+
+    finalWeights
+  }
+
+  /**
+    * CostFun implements Breeze's DiffFunction[T], which returns the loss and gradient
+    * at a particular point (weights). It's used in Breeze's convex optimization routines.
+    */
+  private class CostFun[T <: Vector[Double]](
+      data: RDD[T],
+      labels: RDD[DenseVector[Double]],
+      gradient: Gradient[T],
+      regParam: Double,
+      numExamples: Long,
+      numFeatures: Int,
+      numClasses: Int) extends DiffFunction[DenseVector[Double]] {
+
+    override def calculate(weights: DenseVector[Double]): (Double, DenseVector[Double]) = {
+      val weightsMat = weights.asDenseMatrix.reshape(numFeatures, numClasses)
+      // Have a local copy to avoid the serialization of CostFun object which is not serializable.
+      val bcW = data.context.broadcast(weightsMat)
+      val localGradient = gradient
+      val localNumFeatures = numFeatures
+      val localNumClasses = numClasses
+
+      val partitionGradientLosses = data.zipPartitions(labels) {
+        (partitionFeatures, partitionLabels) =>
+          Iterator.single(localGradient.compute(localNumFeatures, localNumClasses, partitionFeatures, partitionLabels, bcW.value))
+      }
+
+      val (gradientSum, lossSum) = MLMatrixUtils.treeReduce(
+        partitionGradientLosses,
+        (a: (DenseMatrix[Double], Double), b: (DenseMatrix[Double], Double)) => {
+          a._1 += b._1
+          (a._1, a._2 + b._2)
+        }
+      )
+
+      // total loss = lossSum / nTrain + 1/2 * lambda * norm(W)^2
+      val normWSquared = math.pow(norm(weights), 2)
+      val regVal = 0.5 * regParam * normWSquared
+      val loss = lossSum / numExamples + regVal
+
+      // total gradient = gradSum / nTrain + lambda * w
+      val gradientTotal = gradientSum / numExamples.toDouble + (weightsMat * regParam)
+
+      (loss, gradientTotal.toDenseVector)
+    }
+  }
+}
+
+/**
+  * :: DeveloperApi ::
+  * Class used to solve an optimization problem using Limited-memory BFGS.
+  * Reference: [[http://en.wikipedia.org/wiki/Limited-memory_BFGS]]
+  * @param gradient Gradient function to be used.
+  * @param numCorrections 3 < numCorrections < 10 is recommended.
+  * @param convergenceTol convergence tolerance for L-BFGS
+  * @param regParam L2 regularization
+  * @param numIterations max number of iterations to run
+  */
+class DenseLBFGSwithL2(
```

Comment:
We can make the data mean scaling an earlier step in the pipeline, but should we leave the label mean scaling here? (we can do that for the sparse case too, although my impression from the tests is it can make things weird if we only mean scale the labels but not the data or vice versa)